### PR TITLE
Refactor http client

### PR DESCRIPTION
**Pull request by shhdgit**
Refactor `httpc` by resty for better abstraction.

### CODE REVIEW

File: pkg/httpc/client.go [hunk 1]
```diff
@@ -35,100 +35,44 @@ const (
...
+			if r.IsError() {
+				err := fmt.Errorf(string(r.Body()))
+				log.Warn("SendRequest failed", zap.Strings("Request", []string{r.Request.Method, r.Request.URL}), zap.Error(err))
+				return err
+			}
+			return nil
+		})
+	c.lifecycleCtx.Append(fx.Hook{
+		OnStop: func(c context.Context) error {
+			client.GetClient().CloseIdleConnections()
```
Comment (shhdgit) [inline pkg/httpc/client.go:69]: I think there are too many client references held in this lifecycle hook.

**Comment by shhdgit** [comment]
PTAL!
/cc @breeswish

File: pkg/apiserver/configuration/service.go [hunk 2]
```diff
@@ -131,11 +131,13 @@ func (s *Service) getConfigItemsFromTiDBToChannel(tidb *topology.TiDBInfo, ch ch
 }
 
 func (s *Service) getConfigItemsFromTiDB(host string, statusPort int) (map[string]interface{}, error) {
-	data, err := s.params.TiDBClient.WithStatusAPIAddress(host, statusPort).SendGetRequest("/config")
+	resp, err := s.params.TiDBClient.NewStatusAPIClientWithEnforceHost(fmt.Sprintf("%s:%d", host, statusPort)).
```
Comment (shhdgit) [inline pkg/apiserver/configuration/service.go:134]: Should I use `NewStatusAPIClientWithEnforceHost` or `SetHostURL`?

File: pkg/apiserver/debugapi/client.go [hunk 3]
```diff
@@ -57,55 +57,76 @@ func SendRequest(client Client, req *endpoint.Request) (*httpc.Response, error)
...
 	Client *tidb.Client
 }
 
-func (impl *tidbImplement) Get(req *endpoint.Request) (*httpc.Response, error) {
-	return impl.Client.
-		WithEnforcedStatusAPIAddress(req.Host, req.Port).
-		WithStatusAPITimeout(defaultTimeout).
-		Get(buildRelativeURI(req.Path, req.Query))
+func (impl *tidbImplement) Get(req *endpoint.Request) (*http.Response, error) {
+	resp, err := impl.Client.NewStatusAPIClientWithEnforceHost(fmt.Sprintf("%s:%d", req.Host, req.Port)).
```
Comment (shhdgit) [inline pkg/apiserver/debugapi/client.go:66]: ditto.

File: pkg/apiserver/profiling/fetcher.go [hunk 4]
```diff
@@ -77,34 +73,46 @@ type tikvFetcher struct {
...
+	return resp.Body(), err
 }
 
 type tidbFetcher struct {
 	client *tidb.Client
 }
 
 func (f *tidbFetcher) fetch(op *fetchOptions) ([]byte, error) {
-	return f.client.WithStatusAPIAddress(op.ip, op.port).WithStatusAPITimeout(maxProfilingTimeout).SendGetRequest(op.path)
+	resp, err := f.client.NewStatusAPIClientWithEnforceHost(fmt.Sprintf("%s:%d", op.ip, op.port)).
```
Comment (shhdgit) [inline pkg/apiserver/profiling/fetcher.go:100]: ditto.

File: pkg/apiserver/debugapi/client.go [hunk 5]
```diff
@@ -57,55 +57,76 @@ func SendRequest(client Client, req *endpoint.Request) (*httpc.Response, error)
...
-	}
-	return fmt.Sprintf("%s?%s", path, query)
-}
-
 type tidbImplement struct {
 	fx.In
 	Client *tidb.Client
 }
 
-func (impl *tidbImplement) Get(req *endpoint.Request) (*httpc.Response, error) {
```
Comment (breeswish) [inline pkg/apiserver/debugapi/client.go:72]: Is it possible to reduce the encapsulation (for example, remove these helper structs)?

File: pkg/apiserver/debugapi/client.go [hunk 6]
```diff
@@ -57,55 +57,76 @@ func SendRequest(client Client, req *endpoint.Request) (*httpc.Response, error)
...
-	return impl.Client.
-		WithEnforcedStatusAPIAddress(req.Host, req.Port).
-		WithStatusAPITimeout(defaultTimeout).
-		Get(buildRelativeURI(req.Path, req.Query))
+func (impl *tidbImplement) Get(req *endpoint.Request) (*http.Response, error) {
+	resp, err := impl.Client.NewStatusAPIClientWithEnforceHost(fmt.Sprintf("%s:%d", req.Host, req.Port)).
+		// prevent closing body by resty. https://github.com/go-resty/resty/issues/369
+		SetDoNotParseResponse(true).
+		SetTimeout(defaultTimeout).
```
Comment (breeswish) [inline pkg/apiserver/debugapi/client.go:69]: Is it possible to make them default? Seems that this can be easily forgotten.

**Review by breeswish — COMMENTED** [review]
This PR only reduce ~70 LOC, not as effective as my expectation :) Could you find more to make things simple?

**Comment by shhdgit** [comment]
OK, let me think it over and try.

**Comment by breeswish** [comment]
I have made some improvements to the HTTP client in my own branch via a more complete refinement. Closing this PR.

**Pull request closed without merging**